File: app/src/main/java/code/name/monkey/retromusic/mvp/presenter/AlbumDetailsPresenter.java
package code.name.monkey.retromusic.mvp.presenter;

import android.support.annotation.NonNull;
import code.name.monkey.retromusic.model.Album;
import code.name.monkey.retromusic.mvp.Presenter;
import code.name.monkey.retromusic.mvp.contract.AlbumDetailsContract;


/**
 * Created by hemanths on 20/08/17.
 */

public class AlbumDetailsPresenter extends Presenter implements AlbumDetailsContract.Presenter {

  @NonNull
  private final int albumId;
  @NonNull
  private AlbumDetailsContract.AlbumDetailsView view;

  public AlbumDetailsPresenter(@NonNull AlbumDetailsContract.AlbumDetailsView view, int albumId) {

    this.view = view;
    this.albumId = albumId;
  }

  @Override
  public void subscribe() {
    loadAlbumSongs(albumId);
  }

  @Override
  public void unsubscribe() {
    disposable.clear();
  }

  @Override
  public void loadAlbumSongs(int albumId) {
    disposable.add(repository.getAlbum(albumId)
        .subscribeOn(schedulerProvider.computation())
        .observeOn(schedulerProvider.ui())
        .doOnSubscribe(disposable1 -> view.loading())
        .subscribe(this::showAlbum,
            throwable -> view.showEmptyView(),
            () -> view.completed()));
  }

  private void showAlbum(Album album) {
    if (album != null) {
      view.showData(album);
    } else {
      view.showEmptyView();
    }
  }
}

File: app/src/main/java/code/name/monkey/retromusic/mvp/presenter/AlbumDetailsPresenter.kt
/*
 * Copyright (c) 2019 Hemanth Savarala.
 *
 * Licensed under the GNU General Public License v3
 *
 * This is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by
 *  the Free Software Foundation either version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */

package code.name.monkey.retromusic.mvp.presenter

import code.name.monkey.retromusic.model.Album
import code.name.monkey.retromusic.model.Artist
import code.name.monkey.retromusic.mvp.Presenter
import code.name.monkey.retromusic.mvp.PresenterImpl
import code.name.monkey.retromusic.providers.interfaces.Repository
import code.name.monkey.retromusic.rest.model.LastFmAlbum
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

/**
 * Created by hemanths on 20/08/17.
 */
interface AlbumDetailsView {

    fun album(album: Album)

    fun complete()

    fun loadArtistImage(artist: Artist)

    fun moreAlbums(albums: List<Album>)

    fun aboutAlbum(lastFmAlbum: LastFmAlbum)
}

interface AlbumDetailsPresenter : Presenter<AlbumDetailsView> {
    fun loadAlbum(albumId: Int)

    fun loadMore(artistId: Int)

    fun aboutAlbum(artist: String, album: String)

    class AlbumDetailsPresenterImpl @Inject constructor(
        private val repository: Repository
    ) : PresenterImpl<AlbumDetailsView>(), AlbumDetailsPresenter, CoroutineScope {

        private val job = Job()
        private lateinit var album: Album

        override fun loadMore(artistId: Int) {
            launch {
                val result = repository.artistById(artistId)
                showArtistImage(result)
            }
        }

        override fun aboutAlbum(artist: String, album: String) {
            launch {
                val result = repository.albumInfo(artist, album)
                view.aboutAlbum(result)
            }
        }

        private fun showArtistImage(artist: Artist) {
            view?.loadArtistImage(artist)

            artist.albums?.filter { it.id != album.id }?.let {
                if (it.isNotEmpty()) view?.moreAlbums(ArrayList(it))
            }
        }

        override fun loadAlbum(albumId: Int) {
            launch {
                val result = repository.albumById(albumId)
                album = result
                view?.album(result)

            }
        }

        override fun detachView() {
            super.detachView()
            job.cancel()
        }

        override val coroutineContext: CoroutineContext
            get() = Dispatchers.IO + job
    }
}